File: 91-decode-ways/decode-ways.cpp
class Solution {
public:
    int recur(string s, int i, int dp[]){
        if(i == 0 && s[i] == '0') return 0;
        if(i == 0 || i < 0) return 1;
        if(dp[i] != -1) return dp[i];
        int sum1 = 0;
        if(s[i] != '0'){
            sum1 = recur(s,i-1,dp);
        }
        int sum2 = 0;
        if(s[i-1] != '0' && stoi(s.substr(i-1,2)) > 9 && stoi(s.substr(i-1,2)) <= 26){
            sum2 = recur(s,i-2,dp);
        }

        return dp[i] = sum1 + sum2;
        
    }
    int numDecodings(string s) {
        int dp[s.length()];
        memset(dp,-1,s.length() * sizeof(int));
        // return recur(s,0,t);
        int n = s.length();
        return recur(s,n-1,dp);
    }
};

File: 91-decode-ways/decode-ways.py
class Solution:
    def numDecodings(self, s: str) -> int:

        n = len(s)
        dp = [-1] * (n+1)
        dp[n] = 1
        dp[n-1] = (int)(1 <= int(s[n-1]) <= 9)

        def recur(i):
            if i == len(s):
                return 1

            if dp[i] != -1:
                return dp[i]
            
            ans1 = ans2 = 0
            if 1 <= int(s[i]) <= 26:
                ans1 = recur(i+1)
            
            if 10 <= int(s[i:i+2]) <= 26:
                ans2 = recur(i+2)
            
            dp[i] = ans1 + ans2
            return dp[i]
        
        return recur(0)
        